tsk/util/Bitlocker/BitlockerParser.cpp

```diff
@@ -0,0 +1,1553 @@
+
+    ssize_t ret_len2 = tsk_img_read(m_img_info, volumeOffsetToRead + m_volumeOffset, (char*)(&data[ret_len]), bytesLeft);
+    if (ret_len2 == 0) {
+        return ret_len;
+    }
+
+    TSK_DADDR_T offset = volumeOffsetToRead;
+    uint8_t* dataPtr = &(data[ret_len]);
+    while (bytesLeft > 0) {
+        decryptSector(offset, dataPtr);
+
+        offset += m_sectorSize;
+        dataPtr += m_sectorSize;
+        bytesLeft -= m_sectorSize;
+    }
+
+    return ret_len + ret_len2;
+}
+
+/**
+* Decrypt the data that was read from the given offset.
+* 
+* @volumeOffset Offset to the data relative to the start of the volume. Expected to be sector-aligned.
+* @data         Data to decrypt. Should have length m_sectorSize. Will hold the decrypted data.
+* 
+* @return 0 on success, -1 on error.
+*/
+int BitlockerParser::decryptSector(TSK_DADDR_T volumeOffset, uint8_t* data) {
+
+    if (!initializationSuccessful()) {
+        writeError("BitlockerParser::decryptSector: BitlockerParser has not been initialized");
+        return -1;
+    }
+
+    // This seems to only work for Windows 7 (and likely earlier). After that it seems like m_encryptedVolumeSize
+    // is set to the full volume size even when encryption was paused partway through.
+    if (volumeOffset >= m_encryptedVolumeSize) {
+        writeDebug("BitlockerParser::decryptSector: Sector is beyond what was encrypted - returning original data. ");
+        writeDebug("BitlockerParser::decryptSector: Data:         " + convertUint64ToString(volumeOffset) + "   " + convertByteArrayToString(data, 16) + "...");
+        return 0;
+    }
+
+    if (isAESCBC(m_encryptionType)) {
+        if (usesDiffuser(m_encryptionType)) {
+            return decryptSectorAESCBC_diffuser(volumeOffset, data);
+        }
+        else {
+            return decryptSectorAESCBC_noDiffuser(volumeOffset, data);
+        }
+    }
+    else if (isAESXTS(m_encryptionType)) {
+        return decryptSectorAESXTS(volumeOffset, data);
+    }
+    else {
+        writeError("BitlockerParser::decryptSector: Encryption method not currently supported - " + convertEncryptionTypeToString(m_encryptionType));
+        return -1;
+    }
+}
+
+/**
+* Decrypt the data that was read from the given offset using AES-CBC with no diffuser (128 or 256 bit)
+*
+* @volumeOffset Offset to the data relative to the start of the volume. Expected to be sector-aligned.
+* @data         Data to decrypt. Will hold the decrypted data.
+*
+* @return 0 on success, -1 on error.
+*/
+int BitlockerParser::decryptSectorAESCBC_noDiffuser(uint64_t offset, uint8_t* data) {
+
+    // Make temporary buffer to copy encrypted data to
+    uint8_t* encryptedData = (uint8_t*)tsk_malloc(m_sectorSize);
+    if (encryptedData == nullptr) {
+        writeError("BitlockerParser::decryptSectorAESCBC_noDiffuser: Error allocating encryptedData");
+        return -1;
+    }
+    memcpy(encryptedData, data, m_sectorSize);
+    memset(data, 0, m_sectorSize);
```

Comment:
We don't know the size of data here at least from this method, would it be safer to pass in the size of data? Should data always be m_sectorSize and if not throw an error?